### ISSUE TITLE
com.microsoft.azure.sdk.iot.device.exceptions.ModuleClientException: java.net.MalformedURLException: unknown protocol: unix

### ISSUE THREAD

**Issue by mkohns** [Azure/azure-iot-sdk-java#274]
Hi, I have strange behavior using the sdk with iot edge.

IOT Edge:
OS: ubuntu 18.04
iotedge version: 1.0.0 (52ef77db24126bf473265fc09c53d35290a2dd6b)
moby version:  1.0.0-1
docker version: 18.06.0-dev, build daf021fe
Installations Instructions: https://docs.microsoft.com/de-de/azure/iot-edge/quickstart-linux

azure-iot-sdk-java:
```    
<dependency>
      <groupId>com.microsoft.azure.sdk.iot</groupId>
      <artifactId>iot-device-client</artifactId>
      <version>1.13.1</version>
</dependency>
```

In my code I initialize my test iot edge module with

`ModuleClient module = ModuleClient.createFromEnvironment(IotHubClientProtocol.MQTT);`

and get the Exception

```
com.microsoft.azure.sdk.iot.device.exceptions.ModuleClientException: java.net.MalformedURLException: unknown protocol: unix
        at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:251)
        at com.microsoft.docs.iothub.samples.SimulatedDevice.main(SimulatedDevice.java:172)
Caused by: java.net.MalformedURLException: unknown protocol: unix
        at java.net.URL.<init>(URL.java:600)
        at java.net.URL.<init>(URL.java:490)
        at java.net.URL.<init>(URL.java:439)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:71)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpHsmSignatureProvider.sign(HttpHsmSignatureProvider.java:84)
        at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:80)
        at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.create(IotHubSasTokenHsmAuthenticationProvider.java:48)
        at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:234)
        ... 1 more
```

docker inspect shows the injected environment variables from iotedge

```
           "Env": [
                "RuntimeLogLevel=Information",
                "IOTEDGE_DEVICEID=edgeDevice1",
                "IOTEDGE_GATEWAYHOSTNAME=ubuntu-bionic",
                "IOTEDGE_APIVERSION=2018-06-28",
                "IOTEDGE_MODULEID=simulated-device-3",
                "IOTEDGE_IOTHUBHOSTNAME=iothubtest1mk.azure-devices.net",
                "IOTEDGE_WORKLOADURI=unix:///var/run/iotedge/workload.sock",
                "IOTEDGE_MODULEGENERATIONID=636672829613278709",
                "IOTEDGE_AUTHSCHEME=sasToken",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
```
please note the IOTEDGE_WORKLOADURI

I actully assumed due to this documentation: https://docs.microsoft.com/de-de/azure/iot-edge/module-development that `EdgeHubConnectionString` is injected into docker environment. This seems not to be the case.
Therefor "else" tree of com.microsoft.azure.sdk.iot.device.ModuleClient(line 170) is evaulated which produced the error if IOTEDGE_WORKLOADURI starts with `unix`

Maybe I am confused of iotedge v1/v2 versions and mixing up sdk´s

**Comment by timtay-microsoft** [comment]
It looks like this is a bug on our side. I'll look into fixing this issue for you. This SDK should support unix addresses for the workload uri after the fix.

**Comment by timtay-microsoft** [comment]
Were you hoping to use the edgehub connection string rather than the HSM path that you are currently using?

**Comment by mkohns** [comment]
That is a good question.
The documentation at https://docs.microsoft.com/de-de/azure/iot-edge/module-development points out that `EdgeHubConnectionString` will be injected into environment.
From my point of view this should be wrong and the socket based connection should be preferred.
But - I am not an expert in iot edge. I am of this opinion to minimize SSL interdependencies

**Comment by timtay-microsoft** [comment]
The IoT sdks will prioritize using an edgehub connection string over an HSM for authentication. If you aren't getting the edgehubconnectionstring injected into your environment, that is probably a bug in the setup. I'll keep digging on the unix socket problem, though.

**Comment by timtay-microsoft** [comment]
I have created a [branch](https://github.com/Azure/azure-iot-sdk-java/tree/unixSocketsFix) in this repo with a fix for your issue. Are you willing to try building the jar files from source and trying out your scenario with this fix?

**Comment by mkohns** [comment]
Yeap! I will test this tonight! I come back to you soon.

**Comment by mkohns** [comment]
The issue is still present .. but stackstrace looks different

```
com.microsoft.azure.sdk.iot.device.exceptions.ModuleClientException: java.io.IOException: java.net.URISyntaxException: Expected closing bracket for IPv6 address at index 13: unix://[unix:/var/run/iotedge/workload.sock/modules/simulated-device-3/genid/636673744337690415/sign?api-version=2018-06-28:-1]/var/run/iotedge/workload.sock/modules/simulated-device-3/genid/636673744337690415/sign?api-version=2018-06-28
        at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:251)
        at com.microsoft.docs.iothub.samples.SimulatedDevice.main(SimulatedDevice.java:175)
Caused by: java.io.IOException: java.net.URISyntaxException: Expected closing bracket for IPv6 address at index 13: unix://[unix:/var/run/iotedge/workload.sock/modules/simulated-device-3/genid/636673744337690415/sign?api-version=2018-06-28:-1]/var/run/iotedge/workload.sock/modules/simulated-device-3/genid/636673744337690415/sign?api-version=2018-06-28
        at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:95)
        at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.create(IotHubSasTokenHsmAuthenticationProvider.java:48)
        at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:234)
        ... 1 more
Caused by: java.net.URISyntaxException: Expected closing bracket for IPv6 address at index 13: unix://[unix:/var/run/iotedge/workload.sock/modules/simulated-device-3/genid/636673744337690415/sign?api-version=2018-06-28:-1]/var/run/iotedge/workload.sock/modules/simulated-device-3/genid/636673744337690415/sign?api-version=2018-06-28
        at java.net.URI$Parser.fail(URI.java:2848)
        at java.net.URI$Parser.failExpecting(URI.java:2854)
        at java.net.URI$Parser.parseServer(URI.java:3231)
        at java.net.URI$Parser.parseAuthority(URI.java:3155)
        at java.net.URI$Parser.parseHierarchical(URI.java:3097)
        at java.net.URI$Parser.parse(URI.java:3053)
        at java.net.URI.<init>(URI.java:673)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsRequestResponseSerializer.serializeRequest(HttpsRequestResponseSerializer.java:62)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendHttpRequestUsingUnixSocket(HttpsHsmClient.java:183)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:111)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpHsmSignatureProvider.sign(HttpHsmSignatureProvider.java:84)
        at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:80)
        ... 3 more
Exception in thread "main" java.lang.NullPointerException
        at com.microsoft.docs.iothub.samples.SimulatedDevice.main(SimulatedDevice.java:181)
```

by the way, I had to disable one more test to mvn package the project.

**Comment by timtay-microsoft** [comment]
I went ahead and made another push to the same branch for you to experiment with. Can you let me know how it goes?

**Comment by JMayrbaeurl** [comment]
Having the same problem (with the same SDK verison) on Ubuntu 18.04. And on Windows with Linux containers a very similar issue: Trying to use an URL prefixed with http for a HTTPS (secured connection).

**Comment by timtay-microsoft** [comment]
@mkohns were you able to use the latest commit I made to make your solution work?

**Comment by timtay-microsoft** [comment]
Just a quick update: I merged the changes from the branch I linked above into a better branch that has a few other fixes. [Here](https://github.com/Azure/azure-iot-sdk-java/tree/tim-e2e) is the link to that branch. I will be deleting the other branch soon.

**Comment by jasmineymlo** [comment]
In https://github.com/Azure/azure-iot-sdk-java/releases/tag/2018-8-10.  Thanks.

**Comment by jmbrunskill** [comment]
We are experiencing a problem that looks identical to this issue even using the latest SDK on Ubuntu 18.04

```
# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.1 LTS
Release:        18.04
Codename:       bionic
```


```
# iotedge version
iotedge 1.0.5 (d76e0316c6f324345d77c48a83ce836d09392699)
```

Error recieved:
```
2019-01-16 21:37:23.144 DEBUG 1 --- [           main] c.f.a.e.m.messagebus.IotEdgeMessageBus   : Start to create client with MQTT protocol
2019-01-16 21:37:23.237 ERROR 1 --- [           main] c.f.a.e.m.messagebus.IotEdgeMessageBus   : Error starting IotEdge Message Bus: java.net.MalformedURLException: unknown protocol: unix
com.microsoft.azure.sdk.iot.device.exceptions.ModuleClientException: java.net.MalformedURLException: unknown protocol: unix
        at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:252)
        at com.fonterra.aothistorian.edgemodule.messageprocessor.messagebus.IotEdgeMessageBus.start(IotEdgeMessageBus.java:92)
        at com.fonterra.aothistorian.edgemodule.messageprocessor.AppReadyEventListener.onApplicationEvent(AppReadyEventListener.java:27)
        at com.fonterra.aothistorian.edgemodule.messageprocessor.AppReadyEventListener.onApplicationEvent(AppReadyEventListener.java:11)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
        at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:398)
        at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:355)
        at org.springframework.boot.context.event.EventPublishingRunListener.running(EventPublishingRunListener.java:105)
        at org.springframework.boot.SpringApplicationRunListeners.running(SpringApplicationRunListeners.java:78)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:332)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248)
        at com.fonterra.aothistorian.edgemodule.messageprocessor.App.main(App.java:12)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.net.MalformedURLException: unknown protocol: unix
        at java.net.URL.<init>(URL.java:600)
        at java.net.URL.<init>(URL.java:490)
        at java.net.URL.<init>(URL.java:439)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:197)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:111)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpHsmSignatureProvider.sign(HttpHsmSignatureProvider.java:85)
        at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:80)
        at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.create(IotHubSasTokenHsmAuthenticationProvider.java:48)
        at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:235)
        ... 22 more

```

Docker Inspect output
```            "Env": [
                "IOTEDGE_IOTHUBHOSTNAME=aot-historian-dev-iot-hub.azure-devices.net",
                "IOTEDGE_GATEWAYHOSTNAME=nzshstiotedg01m",
                "DATAITEM_IDENTIFIER_PATTERN=NZ\\d\\d\\d.*",
                "IOTEDGE_WORKLOADURI=unix:///var/run/iotedge/workload.sock",
                "IOTEDGE_MODULEGENERATIONID=636831035590658786",
                "SITE_CODE=NZ060",
                "IOTEDGE_DEVICEID=NZSHSTIOTEDG01M",
                "IOTEDGE_AUTHSCHEME=sasToken",
                "IOTEDGE_APIVERSION=2018-06-28",
                "RuntimeLogLevel=Debug",
                "IOTEDGE_MODULEID=messageprocessor4j",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/lib/jvm/java-1.8-openjdk/jre/bin:/usr/lib/jvm/java-1.8-openjdk/bin",
                "LANG=C.UTF-8",
                "JAVA_HOME=/usr/lib/jvm/java-1.8-openjdk",
                "JAVA_VERSION=8u191",
                "JAVA_ALPINE_VERSION=8.191.12-r0"
            ],
```

POM File Dependencies:

```
   <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <java.version>1.8</java.version>
        <azure.iot.device.client.version>1.14.2</azure.iot.device.client.version>
    </properties>
    <dependencies>
        <dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-actuator</artifactId>
        </dependency>
        <dependency>
            <groupId>io.micrometer</groupId>
            <artifactId>micrometer-registry-jmx</artifactId>
            <version>1.1.1</version>
        </dependency>
        <dependency>
            <groupId>com.microsoft.azure.sdk.iot</groupId>
            <artifactId>iot-device-client</artifactId>
            <version>${azure.iot.device.client.version}</version>
        </dependency>
 </dependencies>
```

Is it possible that the spring framework would be interfering with this fix?

**Comment by timtay-microsoft** [comment]
@jmbrunskill I recommend double checking your dependency version on this SDK. The stack trace you provided is consistent only with an older version of our SDK. Note the line number on your stack trace 
 com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:197)

points to an empty statement line on the latest SDK. 
```java
{ 
```

A few versions back, it pointed to this line:

```java
requestUrl = new URL(baseUri + path + "?" + queryString);
```

which is where I expect to see that "unknown protocol: XXXXX" exception.

**Comment by jmbrunskill** [comment]
When building and deploy the sample java module, the client connection works fine. So there must be something else happening in our code, or a spring boot interaction issue. I'll keep investigating and document what I find here in case someone else runs into the same issue.

**Comment by jmbrunskill** [comment]
@timtay-microsoft It looks like this issue isn't triggered by spring-boot exactly but by the embedded tomcat server which is the default for a spring-boot-starter app.

The tomcat code registers a custom URLStreamHandlerFactory before the IoTEdge Library tries to register it's own custom HandlerFactory.

In this scenario, the setUrlStreamHandlerFactory function throws a a 'Factory already defined' error.
However the edge module simply ignores this error making it difficult to identify without interactive debugging.

For our purposes we were able to work around this issue by using spring-boot-starter-webflux.

If anyone else arrives here experiencing a similar issue, this involves simply changing your dependency 
```
<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-web</artifactId>
</dependency>
```
to 
```
<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-webflux</artifactId>
</dependency>
```
This will cause your code to start using Netty instead of tomcat and at this stage Netty doesn't appear to attempt to register a UrlStreamHandlerFactory.

This gives us a work around for now, meaning the fix isn't critical to our use case.

However, I'm wondering if I should I raise a new issue for this problem?
I feel the library should at least throw an exception in the case where the HandlerFactory can't be created to assist with debugging if anyone else experiences this type of issue in future.

Relevant code appears to be in HttpsHsmClient.java		
```
try
            {
				URL.setURLStreamHandlerFactory(fac);	
			}
			catch (Error e)
            {
				//this function only throws if the factory has already been set, so we can ignore this error
			}
```

Perhaps a simple check to see if the Registered UrlStreamHandlerFactory is of the correct class if an error is thrown here?

I've created a github demo of this issue if you want to have some code to test out: https://github.com/jmbrunskill/azure-iot-sdk-java-spring-demo

**Comment by timtay-microsoft** [comment]
Thanks for doing the investigation on this issue. If you'd like to submit a PR to add some logging/exception throwing for this scenario, I'd be happy to review it and merge it in for you.

**Comment by Partha17** [comment]
> We are experiencing a problem that looks identical to this issue even using the latest SDK on Ubuntu 18.04
> 
> ```
> # lsb_release -a
> No LSB modules are available.
> Distributor ID: Ubuntu
> Description:    Ubuntu 18.04.1 LTS
> Release:        18.04
> Codename:       bionic
> ```
> 
> ```
> # iotedge version
> iotedge 1.0.5 (d76e0316c6f324345d77c48a83ce836d09392699)
> ```
> 
> Error recieved:
> 
> ```
> 2019-01-16 21:37:23.144 DEBUG 1 --- [           main] c.f.a.e.m.messagebus.IotEdgeMessageBus   : Start to create client with MQTT protocol
> 2019-01-16 21:37:23.237 ERROR 1 --- [           main] c.f.a.e.m.messagebus.IotEdgeMessageBus   : Error starting IotEdge Message Bus: java.net.MalformedURLException: unknown protocol: unix
> com.microsoft.azure.sdk.iot.device.exceptions.ModuleClientException: java.net.MalformedURLException: unknown protocol: unix
>         at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:252)
>         at com.fonterra.aothistorian.edgemodule.messageprocessor.messagebus.IotEdgeMessageBus.start(IotEdgeMessageBus.java:92)
>         at com.fonterra.aothistorian.edgemodule.messageprocessor.AppReadyEventListener.onApplicationEvent(AppReadyEventListener.java:27)
>         at com.fonterra.aothistorian.edgemodule.messageprocessor.AppReadyEventListener.onApplicationEvent(AppReadyEventListener.java:11)
>         at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
>         at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
>         at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
>         at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:398)
>         at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:355)
>         at org.springframework.boot.context.event.EventPublishingRunListener.running(EventPublishingRunListener.java:105)
>         at org.springframework.boot.SpringApplicationRunListeners.running(SpringApplicationRunListeners.java:78)
>         at org.springframework.boot.SpringApplication.run(SpringApplication.java:332)
>         at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260)
>         at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248)
>         at com.fonterra.aothistorian.edgemodule.messageprocessor.App.main(App.java:12)
>         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>         at java.lang.reflect.Method.invoke(Method.java:498)
>         at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
>         at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
>         at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
>         at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
> Caused by: java.net.MalformedURLException: unknown protocol: unix
>         at java.net.URL.<init>(URL.java:600)
>         at java.net.URL.<init>(URL.java:490)
>         at java.net.URL.<init>(URL.java:439)
>         at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:197)
>         at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:111)
>         at com.microsoft.azure.sdk.iot.device.hsm.HttpHsmSignatureProvider.sign(HttpHsmSignatureProvider.java:85)
>         at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:80)
>         at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.create(IotHubSasTokenHsmAuthenticationProvider.java:48)
>         at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:235)
>         ... 22 more
> ```
> 
> Docker Inspect output
> 
> ```
>                 "IOTEDGE_IOTHUBHOSTNAME=aot-historian-dev-iot-hub.azure-devices.net",
>                 "IOTEDGE_GATEWAYHOSTNAME=nzshstiotedg01m",
>                 "DATAITEM_IDENTIFIER_PATTERN=NZ\\d\\d\\d.*",
>                 "IOTEDGE_WORKLOADURI=unix:///var/run/iotedge/workload.sock",
>                 "IOTEDGE_MODULEGENERATIONID=636831035590658786",
>                 "SITE_CODE=NZ060",
>                 "IOTEDGE_DEVICEID=NZSHSTIOTEDG01M",
>                 "IOTEDGE_AUTHSCHEME=sasToken",
>                 "IOTEDGE_APIVERSION=2018-06-28",
>                 "RuntimeLogLevel=Debug",
>                 "IOTEDGE_MODULEID=messageprocessor4j",
>                 "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/lib/jvm/java-1.8-openjdk/jre/bin:/usr/lib/jvm/java-1.8-openjdk/bin",
>                 "LANG=C.UTF-8",
>                 "JAVA_HOME=/usr/lib/jvm/java-1.8-openjdk",
>                 "JAVA_VERSION=8u191",
>                 "JAVA_ALPINE_VERSION=8.191.12-r0"
>             ],
> ```
> 
> POM File Dependencies:
> 
> ```
>    <properties>
>         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
>         <java.version>1.8</java.version>
>         <azure.iot.device.client.version>1.14.2</azure.iot.device.client.version>
>     </properties>
>     <dependencies>
>         <dependency>
> 			<groupId>org.springframework.boot</groupId>
> 			<artifactId>spring-boot-starter-web</artifactId>
> 		</dependency>
>         <dependency>
>             <groupId>org.springframework.boot</groupId>
>             <artifactId>spring-boot-starter-actuator</artifactId>
>         </dependency>
>         <dependency>
>             <groupId>io.micrometer</groupId>
>             <artifactId>micrometer-registry-jmx</artifactId>
>             <version>1.1.1</version>
>         </dependency>
>         <dependency>
>             <groupId>com.microsoft.azure.sdk.iot</groupId>
>             <artifactId>iot-device-client</artifactId>
>             <version>${azure.iot.device.client.version}</version>
>         </dependency>
>  </dependencies>
> ```
> 
> Is it possible that the spring framework would be interfering with this fix?
Is this issue fixed? Please comment . I am facing the exact same issue.

**Comment by jmbrunskill** [comment]
@Partha17 I worked around this issue by changing my spring boot dependancy to use webflux

```
<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-webflux</artifactId>
</dependency>
```

Is that an option for you? This will use an embedded netty server instead of Tomcat.

The spring-boot-starter-web dependancy uses Tomcat which also overrides the setURLStreamHandlerFactory in Java.

**Comment by Partha17** [comment]
@jmbrunskill I tried that. Since I am migrating from a spring boot app, there are hard dependency on spring-boot-starter-web I realised.
I have overcome the issue though...
If I inject IoTHubConnectionString through set modules, the module doesn't search for connection string hence doesn't go to unix url. (Module Client class) 
I think hsm will work for windows machine, I will try and update tomorrow.

Hoping a fix would come for starter-web too..

**Comment by timtay-microsoft** [comment]
Re-opening this thread because we plan to fix this issue altogether by removing the troublesome line altogether

```java
URL.setURLStreamHandlerFactory(fac);	
```

**Comment by timtay-microsoft** [comment]
Closing this issue since the bug has been fixed since 1.19.1. ModuleClient no longer makes any calls to 
```
URL.setURLStreamHandlerFactory(fac);	
```

**Comment by az-iot-builder-01** [comment]
@mkohns, @JMayrbaeurl, @jmbrunskill, @Partha17, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
